File: scripts/hooks/pre_commit_check.py
#!/usr/bin/env python3
import re
import subprocess
import sys

FORBIDDEN_PATTERNS = [
    # OpenAI-style secret (require longer length and ensure not part of a word)
    r"(?<![A-Za-z0-9_-])sk-[A-Za-z0-9_-]{24,}",
    # AWS access key id
    r"(?<![A-Za-z0-9_-])AKIA[0-9A-Z]{16,}",
    # GitHub personal token prefix
    r"(?<![A-Za-z0-9_-])ghp_[A-Za-z0-9_\-]{10,}",
    # PEM private key header
    r"-----BEGIN PRIVATE KEY-----",
]

COMPILED = re.compile("|".join(f"(?:{p})" for p in FORBIDDEN_PATTERNS))

# Paths to ignore (examples, docs, or intentionally placeholder files)
IGNORE_PATH_PREFIXES = (
    '.env.example',
    'docker/secrets/',
    'docs/',
    'README.md',
    '.kiro/',
)

IGNORE_LINE_KEYWORDS = (
    'your_',
    'REPLACED_',
    'placeholder',
    'example',
    '#',
)

# Extend ignores for tests, tools, and common placeholder keywords
IGNORE_PATH_PREFIXES = IGNORE_PATH_PREFIXES + (
    'services/trading-agents/tests/',
    'tools/',
    'services/zep_graphiti/',
)

# Ignore node_modules and nested package example files
IGNORE_PATH_PREFIXES = IGNORE_PATH_PREFIXES + (
    'node_modules/',
    'services/trading-agents/node_modules/',
)

# Also ignore our own scripts/hooks so the hook doesn't flag its own patterns
IGNORE_PATH_PREFIXES = IGNORE_PATH_PREFIXES + (
    'scripts/',
    'scripts/hooks/',
)

IGNORE_LINE_KEYWORDS = IGNORE_LINE_KEYWORDS + (
    'fake',
    'simulated',
    'sk-fake',
    'sk-development',
    'sk-test',
    'TEST',
    'test placeholder',
    'REDACTED',
)


def get_staged_files():
    # Allow pre-commit to pass files as arguments; if provided, use them
    if len(sys.argv) > 1:
        return [p for p in sys.argv[1:] if p]

    out = subprocess.run(["git", "diff", "--cached", "--name-only", "--diff-filter=ACM"], capture_output=True, text=True)
    if out.returncode != 0:
        print("[pre-commit] Failed to list staged files", file=sys.stderr)
        sys.exit(2)
    files = [x.strip() for x in out.stdout.splitlines() if x.strip()]
    return files


def file_blob_content(path):
    # Read file content from index (staged blob)
    out = subprocess.run(["git", "show", f":{path}"], capture_output=True)
    if out.returncode != 0:
        return None
    try:
        return out.stdout.decode('utf-8', errors='replace')
    except Exception:
        return out.stdout.decode('latin-1', errors='replace')


def main():
    files = get_staged_files()
    if not files:
        return 0

    failed = False
    for f in files:
        # skip ignored paths
        if any(f.startswith(p) for p in IGNORE_PATH_PREFIXES):
            continue
        try:
            # verify file exists in index
            ok = subprocess.run(["git", "ls-files", "--error-unmatch", "--", f], capture_output=True)
            if ok.returncode != 0:
                continue
        except Exception:
            continue

        content = file_blob_content(f)
        if not content:
            continue

        for i, line in enumerate(content.splitlines(), start=1):
            if COMPILED.search(line):
                # skip obvious placeholder lines
                if any(k in line for k in IGNORE_LINE_KEYWORDS):
                    continue
                if not failed:
                    print("[pre-commit] Forbidden token(s) found in staged files:")
                print(f"  {f}:{i}: {line}")
                failed = True

    if failed:
        print('\n[pre-commit] Commit aborted: remove or rotate secrets and retry.', file=sys.stderr)
        return 1

    return 0


if __name__ == '__main__':
    sys.exit(main())

File: scripts/hooks/pre_commit_check.ts
#!/usr/bin/env node
/*
 TypeScript pre-commit hook to scan staged files for secret-like patterns.
 Runs without /bin/sh and without Python virtualenvs.
*/

import { spawnSync } from 'node:child_process';
import fs from 'node:fs';

type Pattern = { name: string; regex: RegExp };

const FORBIDDEN_PATTERNS: Pattern[] = [
  { name: 'OpenAI-style key', regex: /(?<![A-Za-z0-9_-])sk-[A-Za-z0-9_-]{20,}/ },
  { name: 'AWS Access Key ID', regex: /(?<![A-Z0-9])AKIA[0-9A-Z]{16}(?![A-Z0-9])/ },
  { name: 'GitHub token', regex: /(?<![A-Za-z0-9])gh[pousr]_[A-Za-z0-9]{20,}/ },
  { name: 'Private key block', regex: /-----BEGIN (?:RSA |EC |OPENSSH )?PRIVATE KEY-----/ },
];

const IGNORE_PATH_PREFIXES = [
  'docs/',
  'node_modules/',
  'tools/',
  'scripts/', // allow scripts themselves
  'docker/secrets/',
  'services/', // tests/docs below are ignored by keyword rules too
  '.kiro/',
  '.github/',
];

const EXPLICIT_IGNORE_FILES = new Set<string>([
  '.env.example',
  '.env.monitoring.example',
]);

const IGNORE_LINE_KEYWORDS = [
  'your_',
  'REPLACED_',
  'fake',
  'simulated',
  'placeholder',
  'example',
  'sk-fake',
];

function isIgnoredPath(path: string): boolean {
  const p = path.replace(/\\/g, '/');
  if (EXPLICIT_IGNORE_FILES.has(p) || EXPLICIT_IGNORE_FILES.has(p.split('/').pop() || '')) return true;
  return IGNORE_PATH_PREFIXES.some(prefix => p.startsWith(prefix));
}

function getStagedFiles(args: string[]): string[] {
  if (args && args.length > 0) return args;
  const res = spawnSync('git', ['diff', '--cached', '--name-only'], { encoding: 'utf-8' });
  if (res.status !== 0) return [];
  return res.stdout.split(/\r?\n/).map(s => s.trim()).filter(Boolean);
}

function readStagedBlob(path: string): string {
  const res = spawnSync('git', ['show', `:${path}`], { encoding: 'utf-8', maxBuffer: 10 * 1024 * 1024 });
  if (res.status !== 0) return '';
  return res.stdout || '';
}

function likelyBinary(content: string): boolean {
  // Simple heuristic: presence of many NULs or very high non-text ratio
  const N = Math.min(content.length, 2048);
  let nonText = 0;
  for (let i = 0; i < N; i++) {
    const c = content.charCodeAt(i);
    if (c === 0) return true;
    if ((c < 9) || (c > 13 && c < 32)) nonText++;
  }
  return nonText > N * 0.3;
}

function lineLooksPlaceholder(line: string): boolean {
  const lower = line.toLowerCase();
  return IGNORE_LINE_KEYWORDS.some(k => lower.includes(k));
}

function main() {
  const args = process.argv.slice(2);
  const files = getStagedFiles(args);

  if (files.length === 0) {
    process.exit(0);
  }

  const findings: { file: string; name: string; match: string }[] = [];

  for (const f of files) {
    if (isIgnoredPath(f)) continue;
    try {
      // Skip large files quickly
      const stat = fs.existsSync(f) ? fs.statSync(f) : undefined;
      if (stat && stat.size > 2 * 1024 * 1024) continue;

      const content = readStagedBlob(f);
      if (!content || likelyBinary(content)) continue;

      const lines = content.split(/\r?\n/);
      for (const pattern of FORBIDDEN_PATTERNS) {
        for (const line of lines) {
          if (lineLooksPlaceholder(line)) continue;
          const m = line.match(pattern.regex);
          if (m) {
            findings.push({ file: f, name: pattern.name, match: m[0].slice(0, 8) + '…' });
            break; // one finding per pattern per file is enough
          }
        }
      }
    } catch {
      // Ignore read errors silently for hook robustness
    }
  }

  if (findings.length > 0) {
    console.error('Pre-commit security checks failed. Potential secrets found:');
    for (const f of findings) {
      console.error(` - ${f.file} (${f.name}) e.g., ${f.match}`);
    }
    process.exit(1);
  }

  process.exit(0);
}

main();
